File: Objective-C/AlgorithmClubObjectiveC/Algorithm/MergeSort.m
//
//  MergeSort.m
//  AlgorithmClubObjectiveC
//
//  Created by Панов Андрей on 11.02.16.
//  Copyright © 2016 Andrey Panov. All rights reserved.
//

#import "MergeSort.h"

@implementation MergeSort

+ (NSArray *)mergeSort:(NSArray *)array {
    
    if (array.count <= 1) return array;
    
    NSUInteger middleIndex = array.count / 2;
    NSArray *leftArray = [MergeSort mergeSort:[array subarrayWithRange:NSMakeRange(0, middleIndex)]];
    NSArray *rightArray = [MergeSort mergeSort:[array subarrayWithRange:NSMakeRange(middleIndex, array.count-middleIndex)]];
    
    return [MergeSort mergeWithLeftPile:leftArray rightPile:rightArray];
}

+ (NSArray *)mergeWithLeftPile:(NSArray *)leftPile rightPile:(NSArray *)rightPile {
    
    NSUInteger leftIndex = 0;
    NSUInteger rightIndex = 0;
    NSMutableArray *orderedArray = [[NSMutableArray alloc] initWithCapacity:leftPile.count + rightPile.count];
    
    while (leftIndex < leftPile.count && rightIndex < rightPile.count) {
        
        if ([leftPile[leftIndex] integerValue] < [rightPile[rightIndex] integerValue]) {
            
            [orderedArray addObject:leftPile[leftIndex]];
            leftIndex++;
        } else if ([leftPile[leftIndex] integerValue] < [rightPile[rightIndex] integerValue]) {
            
            [orderedArray addObject:rightPile[rightIndex]];
            rightIndex++;
        } else {
            
            [orderedArray addObject:leftPile[leftIndex]];
            leftIndex++;
            [orderedArray addObject:rightPile[rightIndex]];
            rightIndex++;
        }
    }
    
    while (leftIndex < leftPile.count) {
        
        [orderedArray addObject:leftPile[leftIndex]];
        leftIndex++;
    }
    
    while (rightIndex < rightPile.count) {
        
        [orderedArray addObject:rightPile[rightIndex]];
        rightIndex++;
    }
    return [NSArray arrayWithArray:orderedArray];
}

@end

File: Swift/AlgorithmClubSwift/Algorithms/MergeSort.swift
//
//  File.swift
//  AlgorithmClubSwift
//
//  Created by Панов Андрей on 15.02.16.
//  Copyright © 2016 Andrey Panov. All rights reserved.
//

import Foundation

func mergeSort(unsortedArray: [Int]) -> [Int] {
    
    guard unsortedArray.count > 1 else { return unsortedArray }
    
    let midIndex = unsortedArray.count / 2
    let leftPile = mergeSort(Array(unsortedArray[0..<midIndex]))
    let rightPile = mergeSort(Array(unsortedArray[midIndex..<unsortedArray.count]))
    
    return mergeSortWithLeftPile(leftPile, rightPile: rightPile)
}

private func mergeSortWithLeftPile(leftPile: [Int], rightPile: [Int]) -> [Int] {
    
    var leftIndex = 0
    var rightIndex = 0
    var array: [Int] = []
    
    while (leftIndex < leftPile.count && rightIndex < rightPile.count) {
        
        if (leftPile[leftIndex] < rightPile[rightIndex]) {
            
            array.append(leftPile[leftIndex])
            leftIndex++
        }
        else if (leftPile[leftIndex] > rightPile[rightIndex]) {
            
            array.append(rightPile[rightIndex])
            rightIndex++
        }
        else {
    
            array.append(leftPile[leftIndex])
            leftIndex++
            array.append(rightPile[rightIndex])
            rightIndex++
        }
    }
    
    while (leftIndex < leftPile.count) {
        
        array.append(leftPile[leftIndex])
        leftIndex++
    }
    while (rightIndex < rightPile.count) {
        
        array.append(rightPile[rightIndex])
        rightIndex++
    }
    return array
}